### PR TITLE
ReadMe to Fern docs converter: convert openapi specs from readme to fern docs

### PR DESCRIPTION
**Pull request by rishan10**
Given a ReadMe docs url, the cli will init all the openapis (one openapi init failing won't block the rest) it can find along with an init of docs. 

## Changes:
- modified init docs to also init an openapi workspace if we pass in a docs url that contains an open api spec
- added docs utils which has an interface for docs url actions (ex: get all openapis in docs)

## Testing:
- e2e test for successful init of openapi docs

### CODE REVIEW

File: packages/cli/cli/src/cli.ts [hunk 1]
```diff
@@ -187,16 +188,31 @@ function addInitCommand(cli: Argv<GlobalCliOptions>, cliContext: CliContext) {
...
                 return cliContext.failWithoutThrowing("Cannot specify both --api and --docs. Please choose one.");
             } else if (argv.docs != null) {
+                let docsUrl: ReadMeDocsUrl;
+                if (argv.url != null) {
+                    docsUrl = new ReadMeDocsUrl(argv.url);
+                    if (!docsUrl.isUrlValid()) {
+                        return cliContext.failWithoutThrowing("--url must be a valid url");
+                    }
+                    if (argv.organization == null) {
+                        return cliContext.failWithoutThrowing("--organization must be specified when using --url");
```
Comment (dsinghvi) [inline packages/cli/cli/src/cli.ts:207]: nit: would be sick if it just prompoted user for organization https://stackoverflow.com/questions/33858763/console-input-in-typescript

File: packages/cli/ete-tests/src/tests/init/init.test.ts [hunk 2]
```diff
@@ -60,4 +60,30 @@ describe("fern init", () => {
...
+                cwd: pathOfDirectory,
+            }
+        );
+
+        const command = await runFernCli(["check"], {
+            cwd: pathOfDirectory,
+        });
+
+        expect(command.exitCode).toBe(0);
+    }, 60_000);
```
Comment (dsinghvi) [inline packages/cli/ete-tests/src/tests/init/init.test.ts:88]: looks great!

File: packages/cli/init/src/initializeDocs.ts [hunk 3]
```diff
@@ -1,41 +1,105 @@
...
+
     return {
         instances: [
             {
                 url: `https://${organization}.${process.env.DOCS_DOMAIN_SUFFIX}`,
             },
         ],
         title: `${organization} | Documentation`,
-        navigation: [{ api: "API Reference" }],
+        navigation: apiSection,
```
Comment (dsinghvi) [inline packages/cli/init/src/initializeDocs.ts:102]: nit: you can inline `getApiSection()` here

File: packages/cli/init/src/utils/docsUrlUtils.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            new URL(this.url);
+            return true;
+        } catch (err) {
+            return false;
+        }
+    }
+}
+
+export class ReadMeDocsUrl extends DocsUrlBase {
+    type: "readme" = "readme";
```
Comment (dsinghvi) [inline packages/cli/init/src/utils/docsUrlUtils.ts:43]: nit: would make this public readonly so it can't be modified

Comment (dsinghvi) [reply packages/cli/init/src/utils/docsUrlUtils.ts:43]: `public readonly type = "readme";`

File: packages/cli/init/src/utils/docsUrlUtils.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+    async getAllOpenApiUrls(): Promise<string[]> {
+        const origin = this.getBaseUrl();
+        const response = await axios.get(origin + "/openapi");
```
Comment (dsinghvi) [inline packages/cli/init/src/utils/docsUrlUtils.ts:47]: i would make sure you are safely joining and the leading `/` does not conflict

File: packages/cli/init/src/utils/docsUrlUtils.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+        if (response.status / 100 !== 2) {
+            return [];
+        }
```
Comment (dsinghvi) [inline packages/cli/init/src/utils/docsUrlUtils.ts:51]: what is this doing?

File: packages/cli/init/src/utils/docsUrlUtils.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+export type DocsUrl = ReadMeDocsUrl;
+
+export interface DocsUrlInterface {
+    getAllOpenApiUrls(): Promise<string[]>;
+    getBaseUrl(): string;
+    isUrlValid(): boolean;
+}
+
+class DocsUrlBase implements DocsUrlInterface {
```
Comment (dsinghvi) [inline packages/cli/init/src/utils/docsUrlUtils.ts:12]: what do you think of changing the naming to: 

```
class DocsInstance implements DocsInterface {

}

class ReadmeInstance extends DocsInstance {

}
```

Comment (rishan10) [reply packages/cli/init/src/utils/docsUrlUtils.ts:12]: what about 
`class DocsImpl implements DocsInterface {

}

class ReadmeImpl extends DocsInstance {

}`? Noticed [impl](https://sourcegraph.com/github.com/fern-api/fern@add-cli-option/-/blob/packages/cli/cli/src/cli-context/TaskContextImpl.ts?L31) is used in the codebase in other places: https://sourcegraph.com/github.com/fern-api/fern@add-cli-option/-/blob/packages/cli/cli/src/cli-context/TaskContextImpl.ts?L31

File: packages/cli/init/src/utils/docsUrlUtils.ts (hunk 5 above)
Comment (rishan10) [reply packages/cli/init/src/utils/docsUrlUtils.ts]: ah yes, accomplished this with a URL builder!

File: packages/cli/init/src/utils/docsUrlUtils.ts (hunk 6 above)
Comment (rishan10) [reply packages/cli/init/src/utils/docsUrlUtils.ts]: Apologies, this was by instinct (handling all non-2XX here). I pushed a commit that should make it more clear: if we don't get a 200, then we log an error and return an empty list, signaling that we were unable to find any openapi urls

File: packages/cli/cli/src/cli.ts (hunk 1 above)
Comment (rishan10) [reply packages/cli/cli/src/cli.ts:207]: for sure, let me cut a ticket for this (will do in a new pr)

Comment (rishan10) [reply packages/cli/cli/src/cli.ts:207]: https://linear.app/fern-api/issue/FER-547/prompt-user-to-input-an-organization-when-running-init-docs-url

**Review by dsinghvi — APPROVED** [review]

**Pull request merged**